broadinstitute/single_cell_portal_core
# Validate metadata Excel drag issues (SCP-3645)

SCP-3645, SCP-3786 This adds checking for where multiple ids are matched to the same label, which is one common symptom of the infamous 'excel drag error'.  

![image](https://user-images.githubusercontent.com/2800795/145633961-63973085-3df9-4238-bb02-c85c535259cc.png)

This also adds checking for when a 'group' column (other than NAME) has more than 200 values for both metadata and cluster files.  This just outputs a 'warning'--we allow the upload since their data may have that many unique values, but want to alert the user in case that is inadvertent.

This also adds checking for required metadata columns.  For now, the columns are hardcoded since they change so infrequently.  As we add more client-side validations, we can just import the convention json schema into our javascript and parse that.
![image](https://user-images.githubusercontent.com/2800795/145634155-71377df5-49d9-48e5-8e18-63f7b95fe468.png)


TO TEST:
1.  In upload wizard, attempt to add the `test/test_data/validations/metadata_drag_error.tsv` file as a metadata file
2. Observe errors and warnings as in the screenshot above 
3. Attempt to add the `test/test_data/validations/metadata_bad_no_ontology.txt` file
4. observe errors listing the missing headers as in the screenshot above
5. run `yarn test`, observe all pass
